### PR TITLE
Bringing laplace-torch to foundation-model era

### PR DESCRIPTION
**Pull request by wiseodd**
Main features of this pull request:

1. Support only doing Laplace on params that require grad. Use case: PEFT (like LoRA) on top of a frozen foundation model. This is more efficient than SubnetLaplace since the latter still computes the full Jacobians.
2. Add support to multiple leading dims in `classification` likelihood. E.g. the logits is `(batch_size, seq_len, n_classes)`. Useful for language modeling and reward modeling.
    1. This PR also contains the `integrate-latest-asdl` changes. I tested it with my ASDL fork (only a couple of light changes to support weight-sharing dim and `ignore_index`; so crucial in language modeling): <https://github.com/wiseodd/asdl/commits/dev/>. Please also check this and let me know what's the most elegant way. 
4. Support Huggingface dataset. The assumption is that `x` is a `UserDict` containing `input_ids`, `attention_mask`, etc., things that are produced by HF dataloader.
5. Add a new likelihood called `reward_modeling` where the `classification` likelihood is used during training and the `regression` likelihood is used during prediction.
6. Add support to `torchmetrics` for gridsearch. The benefit is that it supports running metrics => less memory overhead (vis-a-vis gathering all the logits first).
7. Add Jacobian computation with `torch.func` (`functorch`, really) as a general Jacobian computation for GLM predictive. Useful for Bayesian optimization/invariance learning where you need to backprop through the variance. Much more elegant than to change ASDL. 

Relevant unit tests are provided. All tests passed; the only ones failed are the old `LowRankLaplace` issues.

### CODE REVIEW

**Comment by edaxberger** [comment]
Thanks Agustinus for this PR, this is very useful! Just took a glance at the parts that I'm familiar with, which looks good to me. I'll defer to Alex and Runa for a more in-depth review due to their better familiarity with the core library.

File: laplace/utils/matrix.py
Comment (aleximmer) [inline laplace/utils/matrix.py:71]: Why this change? I remember there was a reason for using `.add`, might have had to do with differentiability.

File: laplace/marglik_training.py
Comment (aleximmer) [inline laplace/marglik_training.py:36]: Maybe make this bool and use `sigma_noise_init` as the fixed setting for it. Otherwise, it's unclear how they overwrite each other.

File: laplace/utils/matrix.py
Comment (aleximmer) [inline laplace/utils/matrix.py:151]: Why the else statement although `Q = torch.eye()` for `ndim == 1`? I think in this case you could make Q above a vector of ones and save the memory?

File: laplace/utils/matrix.py
Comment (aleximmer) [inline laplace/utils/matrix.py:392]: plaese remove debug print statements here and below

**Review by aleximmer — COMMENTED** [review]
Thanks, looks overall good to me.

File: laplace/baselaplace.py
Comment (runame) [inline laplace/baselaplace.py:72]: `sum(p.numel() for p in self.params)` is a bit more efficient.

File: laplace/baselaplace.py
Comment (runame) [inline laplace/baselaplace.py:173]: can use `p.numel()` here

File: laplace/baselaplace.py
Comment (runame) [inline laplace/baselaplace.py:650]: Why is this commented out?

File: laplace/curvature/asdl.py
Comment (runame) [inline laplace/curvature/asdl.py:147]: This is against the convention used for `CrossEntropyLoss` in PyTorch, since it expects input shape `(N, C, d_1​, d_2 ​,..., d_K​)` (see the [docs](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)). Could lead to unexpected behaviour for users. If we go with this convention, we should comment on it in the docs and the README, as a requirement for the model.

File: laplace/curvature/asdl.py
Comment (runame) [inline laplace/curvature/asdl.py:179]: same here

File: laplace/curvature/asdl.py
Comment (runame) [inline laplace/curvature/asdl.py:231]: same here

File: laplace/curvature/asdl.py
Comment (runame) [inline laplace/curvature/asdl.py:224]: same here

File: laplace/curvature/asdl.py
Comment (runame) [inline laplace/curvature/asdl.py:173]: same here

File: laplace/curvature/asdl.py
Comment (runame) [inline laplace/curvature/asdl.py:141]: same here

File: laplace/curvature/backpack.py
Comment (runame) [inline laplace/curvature/backpack.py:133]: same here

File: laplace/curvature/backpack.py
Comment (runame) [inline laplace/curvature/backpack.py:148]: same here

File: laplace/curvature/backpack.py
Comment (runame) [inline laplace/curvature/backpack.py:166]: ok I will stop

File: laplace/curvature/curvature.py
Comment (runame) [inline laplace/curvature/curvature.py:105]: Can we avoid this additional forward pass by something like this (from the docs)?
```python
from torch.func import jacrev
x = torch.randn(5)
def f(x):
  return x.sin()
def g(x):
  result = f(x)
  return result, result
jacobian_f, f_x = jacrev(g, has_aux=True)(x)
assert torch.allclose(f_x, f(x))
```

File: laplace/baselaplace.py
Comment (runame) [inline laplace/baselaplace.py:453]: Should there still be this here?
```python
try:
     out = self.model(X[:1].to(self._device))
except (TypeError, AttributeError):
```
It is still there for `LLLaplace`.

File: laplace/lllaplace.py
Comment (runame) [inline laplace/lllaplace.py:161]: Maybe we should do this now, since it will make predictions a lot faster for large foundation models. Should be possible to implement with `forward_with_features()`?

File: laplace/baselaplace.py
Comment (runame) [inline laplace/baselaplace.py:263]: Should we change the default to the `RunningNLLMetric`?

File: laplace/utils/metrics.py
Comment (runame) [inline laplace/utils/metrics.py:8]: ...

File: laplace/utils/metrics.py
Comment (runame) [inline laplace/utils/metrics.py:31]: Against the PyTorch convention again.

File: tests/test_baselaplace.py
Comment (runame) [inline tests/test_baselaplace.py:457]: I assume for debugging?

File: tests/test_curv_backends_asdl.py
Comment (runame) [inline tests/test_curv_backends_asdl.py:129]: remove debugging print statements and comment here and below

File: tests/test_curv_backends_asdl.py
Comment (runame) [inline tests/test_curv_backends_asdl.py:206]: here as well

**Review by runame — CHANGES_REQUESTED** [review]
Thanks a lot for pushing this! Overall looks good to me, see my comments.

File: laplace/utils/matrix.py
Comment (wiseodd) [reply laplace/utils/matrix.py:151]: Again, it's @runame change.

File: laplace/baselaplace.py
Comment (wiseodd) [reply laplace/baselaplace.py:306]: Let's leave this for a new PR

File: laplace/utils/metrics.py
Comment (wiseodd) [reply laplace/utils/metrics.py:8]: As above

File: laplace/curvature/asdl.py
Comment (wiseodd) [reply laplace/curvature/asdl.py:147]: It's from ASDL: https://github.com/kazukiosawa/asdl/blob/9ba6a54bfa4cf3fb21dbc2922f6ef128a3b2e0f9/asdl/fisher.py#L456-L457

Does you suggest we also modify ASDL?

Comment (wiseodd) [reply laplace/curvature/asdl.py:147]: After some readings, I think the current implementation is fine. The input of the model is `N, d_1, ..., d_K, D` where `d_1, ..., d_K` are additional "batch" dimensions and `D` is the input-space dim. Then the model will always output `N, d_1, ..., d_K, C` where `C` is the number of classes.

In Huggingface, they always assume that the logits and labels are flattended, e.g. here for [GPT2](https://github.com/huggingface/transformers/blob/b109257f4fb8b1166e7c53cc5418632014ed53a5/src/transformers/models/gpt2/modeling_gpt2.py#L1336-L1338) and [LLAMA-2](https://github.com/huggingface/transformers/blob/b109257f4fb8b1166e7c53cc5418632014ed53a5/src/transformers/models/llama/modeling_llama.py#L1218-L1224). The loss is then oblivious and agnostic to the fact that we have additional batch dimensions. This is also the current assumption of Laplace and ASDL.

So, I believe as long as we consistently follow the flatten-and-compute-loss convention, we're clear. The user won't be affected at all since it's an internal implementation.

Comment (wiseodd) [reply laplace/curvature/asdl.py:147]: Looking forward to hear your thoughts, though, since I might be wrong!

File: laplace/lllaplace.py
Comment (wiseodd) [reply laplace/lllaplace.py:185]: Seems to be nontrivial at first glance. Let's do this in a separate PR.

File: laplace/utils/matrix.py
Comment (runame) [reply laplace/utils/matrix.py]: I think this is due to changing the initialization of the Kronecker factors to not assume a specific shape by initializing with `0.0` scalars, i.e. not a tensor so `.add` would not work anymore. The motivation for this was to support diagonal Kronecker factors for large layers like embedding layers in language models. See the commit [here](https://github.com/aleximmer/Laplace/pull/144/commits/98a9800db0a28c023ba73444be3cd8d490eeca62).

Comment (runame) [reply laplace/utils/matrix.py]: Not a big fan of this design, so not sure if it should be merged into `main`. Also, regarding @aleximmer's comment, are there tests for checking if the differentiability is affected?

File: laplace/utils/matrix.py
Comment (runame) [reply laplace/utils/matrix.py:154]: `Q.ndim` \neq `Hi.ndim`, but I don't see any case where `Q.ndim == 1` anymore. `Q` is not a vector for diagonal Kronecker factors, because it is not necessarily the identity but some permutation matrix. The code should probably be rewritten to avoid explicitly building the permutation matrix in memory, but this requires duplicating this code in the if-else statement to deal with the diagonal Kronecker factor case separately. Btw, the PR with the changes is [here](https://github.com/aleximmer/Laplace/pull/136).

**Comment by wiseodd** [comment]
Regarding the addition of `logit_class_idx` argument: I did a "smoke test" in implementing this, and realized that the changes would be substantial---even without the present PR, `laplace-torch` always assumes that `logit_class_idx == -1`. 

So, I think this is better done in a separate PR. This will also make the scope of the PRs cleaner---in this PR, we focus on HF LLM models.

**Comment by wiseodd** [comment]
So, the remaining TODOs for this PR are:

1. Decide the conversations about `matrix.py`. @runame & @aleximmer could you please handle this?
2. Wait until https://github.com/f-dangel/curvlinops/pull/100 is merged

Then we can merge this PR and continue with further PRs that we deferred in this PR.

**Comment by runame** [comment]
Sounds good, can you create issues for all comments that you want to address in future PRs?

**Comment by wiseodd** [comment]
I replaced `get_nll` in crossval with `RunningNLLMetrics()`. So this PR will close #160 .
(I thought before that I `RunningNLLMetics` hadn't been implemented, so would have involve major work.)

File: README.md
Comment (aleximmer) [inline README.md:34]: Can you make more precise for what kind of compatibility? This is presumably necessary only for LLM stuff? We should try to merge these changes into ASDL if possible.

File: laplace/utils/matrix.py
Comment (wiseodd) [reply laplace/utils/matrix.py]: I take the point of not assuming a specific shape. I think in the future this is still necessary, but let's build it from the ground up for the next release (0.3). Perhaps along with the addition of `SINGD` backend.

For the 0.2 milestone, let's revert this. (This change also breaks https://github.com/aleximmer/Laplace/issues/171 BTW)

File: laplace/utils/matrix.py
Comment (wiseodd) [reply laplace/utils/matrix.py:166]: Let's leave this be for now then. We can revisit this when we add SINGD baseline. Might be useful then.

File: README.md
Comment (wiseodd) [reply README.md:34]: The PR is open. I think this is non-blocking. We can update the `setup.cfg` once Kazuki merged it.

Comment (wiseodd) [reply README.md:34]: https://github.com/kazukiosawa/asdl/pull/52

**Comment by wiseodd** [comment]
All tasks finished!

**Comment by wiseodd** [comment]
Double-checked and everything looks fine! Merging this.

**Pull request merged**
